pkg/sentry/fsimpl/fuse/read_write.go

```diff
@@ -0,0 +1,152 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package fuse
+
+import (
+	"math"
+	"sync/atomic"
+
+	"gvisor.dev/gvisor/pkg/abi/linux"
+	"gvisor.dev/gvisor/pkg/context"
+	"gvisor.dev/gvisor/pkg/log"
+	"gvisor.dev/gvisor/pkg/sentry/kernel"
+	"gvisor.dev/gvisor/pkg/sentry/kernel/auth"
+	"gvisor.dev/gvisor/pkg/syserror"
+	"gvisor.dev/gvisor/pkg/usermem"
+)
+
+// Read sends a FUSE_READ request, block on it for reply, process the reply and return the payload as a byte slice.
+func (fs *filesystem) Read(ctx context.Context, fd *regularFileFD, off uint64, size uint32) ([]byte, uint32, error) {
+	attributeVersion := atomic.LoadUint64(&fs.conn.attributeVersion)
+
+	// Round up to a multiple of pages size.
+	// size will not be zero, as checked in regularFdReadWriter.ReadToBlocks().
+	readSize, _ := usermem.PageRoundUp(uint64(size))
+
+	// One request cannnot exceed either maxRead or maxPages.
+	maxPages := uint32(math.Ceil(float64(fs.conn.maxRead) / usermem.PageSize))
+	if maxPages > uint32(fs.conn.maxPages) {
+		maxPages = uint32(fs.conn.maxPages)
+	}
+
+	t := kernel.TaskFromContext(ctx)
+	if t == nil {
+		log.Warningf("fusefs.Read: couldn't get kernel task from context")
+		return nil, 0, syserror.EINVAL
+	}
+
+	var outs [][]byte
+	var sizeRead uint32
+
+	// readSize is a multiple of usermem.PageSize.
+	// Always request bytes as a multiple of pages.
+	pagesRead, pagesToRead := uint32(0), uint32(readSize/usermem.PageSize)
+
+	// This loop is intended for fragmented read where the bytes to read is
+	// larger than either the maxPages or maxRead.
+	// For the majority of reads with normal size, this loop should only
+	// execute once.
+	for pagesRead < pagesToRead {
+		pagesCanRead := pagesToRead - pagesRead
+		if pagesCanRead > maxPages {
+			pagesCanRead = maxPages
+		}
+
+		in := linux.FUSEReadIn{
+			Fh:        fd.Fh,
+			Offset:    off + (uint64(pagesRead) << usermem.PageShift),
+			Size:      pagesCanRead << usermem.PageShift,
+			LockOwner: 0, // TODO(gvisor.dev/issue/3245): file lock
+			ReadFlags: 0, // TODO(gvisor.dev/issue/3245): |= linux.FUSE_READ_LOCKOWNER
+			Flags:     fd.statusFlags(),
+		}
+
+		req, err := fs.conn.NewRequest(auth.CredentialsFromContext(ctx), uint32(kernel.TaskFromContext(ctx).ThreadID()), fd.inode().NodeID, linux.FUSE_READ, &in)
+		if err != nil {
+			return nil, 0, err
+		}
+
+		// TODO(gvisor/dev/issue/3247): support async read.
+
+		res, err := fs.conn.Call(t, req)
+		if err != nil {
+			return nil, 0, err
+		}
+		if err := res.Error(); err != nil {
+			return nil, 0, err
+		}
+
+		// No bytes in response, e.g. reached EOF.
+		if len(res.data) <= res.hdr.SizeBytes() {
+			break
+		}
+
+		// Directly using the slice to avoid extra copy.
+		out := res.data[res.hdr.SizeBytes():]
+		outs = append(outs, out)
+		sizeRead += uint32(len(out))
+
+		pagesRead += pagesCanRead
+	}
+
+	// No bytes returned; perhaps user tries to read beyond EOF.
```

Comment:
`offset is beyond EOF` is technically impossible since we just updated the attribute from FUSE server, unless there is a race that shrieked the file. I suppose it's fine to return 0 as in the `man 2 read` page to indicate EOF.

But the convention may be different for gVisor. Need further confirmation. If a `io.EOF` must be returned, it should be returned by the caller.

Reply:
It's not an error. Based on `pkg/sentry/syscalls/linux/error.go:handleIOErrorImpl`, io.EOF can be safely returned.